### ISSUE TITLE
Dead Store rule (1854) should handle destructing assignment with rest property

### ISSUE THREAD

**Issue by jpikl** [SonarSource/SonarJS#994]
**SonarLint version:** 3.5.0.1687 (used with Intellij plugin 3.4.2.2586)
**Rule key:** S1854

**Reproducer**

```
const {foo, ...objectWithoutFoo} = object;
doSomething(objectWithoutFoo);
```

**Expected behavior**

The `foo` variable should not be reported as useless assignment, because it is declared with intention to create shallow copy of `object` without the `foo` property.

Basically, the rule should not apply whenever there is destructing assignment with rest property that is later used in the code. If the `objectWithoutFoo` variable from the previous example was unused, the `foo` and `objectWithoutFoo` would be both reported as useless assignment.
